certora/spec/yieldBox.spec

```diff
@@ -75,369 +60,588 @@ methods {
-                        && f.selector != decimals(uint256).selector  }
-    
-    {
-        preserved with (env e2) {
-            require getAssetsLength() < 1000000;
-        }
-    }  
+hook Sload uint256 amount totalSupply[KEY uint256 id] STORAGE {
+    require to_mathint(amount) == totalSupplyGhost[id];
+}
+
+hook Sstore totalSupply[KEY uint256 id] uint256 amount (uint256 old_amount) STORAGE {
+    totalSupplyGhost[id] = totalSupplyGhost[id] + amount - old_amount;
+}
+
+
+////////////////////////////////////////////////////////////////////////////
+//                       Invariants                                       //
+////////////////////////////////////////////////////////////////////////////
+
+
+/// @title mapArrayCorrealtion
+/// @notice If one of the asset parameters is different then assetId different
+/// @dev How can I specify safe assumptions, which are in the form of requirements, that I want to mention in the report?
+/// @param i One of assetIds to check
+/// @param j One of assetIds to check
+// timeout even with simple code: https://vaas-stg.certora.com/output/3106/4e7e348d8b6741bba4a050f20974f5bd/?anonymousKey=d84620f3b90e2b1fdb78b8d08fb0bf6ff8b042c3 
+// invariant mapArrayCorrealtion(uint i, uint j, env e)
+//     ((i < getAssetsLength() && j < getAssetsLength()) => (assetsIdentical(i, j) <=> i == j))
+//         && (i < getAssetsLength() => ids(e, getAssetTokenType(i), getAssetAddress(i), getAssetStrategy(i), getAssetTokenId(i)) == i)
+//         && (j < getAssetsLength() => ids(e, getAssetTokenType(j), getAssetAddress(j), getAssetStrategy(j), getAssetTokenId(j)) == j)
+
+//     filtered { f -> excludeMethodsDeposit(f)  }
 
+//     {
+//         preserved {
+//             require getAssetsLength() < 1000000;
+//             require i > 0 && j > 0; // safeAssumption because 0 is initialized
+//         }
+//     }
 
-// STATUS - in progress 
-// fails in instate: https://vaas-stg.certora.com/output/3106/b310844c931df786358c/?anonymousKey=d295f27fb55d837678a4cc823b689e25bffbd90c
-invariant idsVsAssets2(YieldData.Asset asset, uint i, uint j, env e)
-        ids(e, getAssetTokenType(i), getAssetAddress(i), getAssetStrategy(i), getAssetTokenId(i)) == i
 
-    filtered { f -> f.selector != batch(bytes[],bool).selector  
-                        && f.selector != uri(uint256).selector 
-                        && f.selector != name(uint256).selector 
-                        && f.selector != symbol(uint256).selector 
-                        && f.selector != decimals(uint256).selector  }
+
+/// @title assetIdtoAssetLength
+/// @notice Existing asset should have id less than assets.length
+/// @param i Asset id
+invariant assetIdtoAssetLength(uint i, env e)
+    ids(e, getAssetTokenType(i), getAssetAddress(i), getAssetStrategy(i), getAssetTokenId(i)) <= getAssetsLength()    
+
+    filtered { f -> excludeMethods(f) }
 
     {
         preserved {
             require getAssetsLength() < 1000000;
```

Comment:
It's just a random number. Does it make any difference if I use MAX_UINT256 instead of a number?

Reply:
just nicer to know that it does not matter. if it does not matter i would you the biggest number i can.

Reply:
it wasn't a random number :) it's necessary to prevent collisions, otherwise, values are changed randomly. 
I added MAX_UINT256 / 4 instead of fixed number. division by 2 still causes collisions.